### PR TITLE
Correctness Testing 2.0

### PR DESCRIPTION
**Pull request by fredjoonpark**
Removed our deprecated correctness testing suite in place of a new approach for correctness testing.

- Added a new class `Mockmetheus` in `./correctness`. It's designed to mock remote operations from Prometheus using snappy-encoded payloads.
- Previously, we had "test" cases that included functions and certain operators that include client-side logic (that stretch beyond remote operations). It now focuses exclusively on the correctness of remote read and write operations.
- Added an optional flag `FRESH_TSDB` for running tests against a freshly new TSDB. The tests are designed to work either way, but defaults to false for convenience.

### Running tests
1. Go to `./correctness` and bring up containers with `docker compose up -d`
2. Run tests with `docker exec -it mockmetheus pytest`

### Introduced test cases for easy reference:
- Remote read on initialization
- Remote read with a DNE metric
- Remote write with an empty timeseries
- Remote write with no labels
- Remote write with no samples
- Remote write & read with valid data
- Remote write & read with multiple metrics
- Remote write & read with multiple samples
- Remote write & read with label matchers

Additional notes:
- `generate_test_run_id()` - helps achieve test case independence by using an extra label in our read/write requests.
- the stub (`prom_pb2.py`) is generated using: `protoc --python_out=. prom.proto`

Mockmetheus (pic for reference):
screenshot_2024-12-18_at_4 55 50___pm_720

### CODE REVIEW

File: correctness/mockmetheus.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+            return write_response.status_code
+        except Exception as e:
+            print(f"Error sending remote write request: {e}")
+            raise
+
+    def parse_duration(self, duration_str):
+        """
+        Parses a Prometheus duration string and returns the duration in milliseconds.
+        """
+        pattern = r'(?:(\d+)h)?(?:(\d+)m)?(?:(\d+)s)?$'
```
Comment (trevorbonas) [inline correctness/mockmetheus.py:116]: Durations can be negative, for example, `-5h`. This wouldn't match the entirety of a negative duration.

Prometheus shows a negative time duration as one possible value in their [documentation](https://prometheus.io/docs/prometheus/latest/querying/basics/#float-literals-and-time-durations):
```
1s # Equivalent to 1.
2m # Equivalent to 120.
1ms # Equivalent to 0.001.
-2h # Equivalent to -7200.
```

File: correctness/mockmetheus.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+            ValueError: If the query string format is invalid.
+        """
+        read_request = ReadRequest()
+
+        # Split the query into main part and hints part if '@' is present
+        parts = query_string.split('@', 1)
+        main_query = parts[0].strip()
+        hints_query = parts[1].strip() if len(parts) > 1 else None
+
+        metric_pattern = r'^\s*([a-zA-Z_:][a-zA-Z0-9_:]*)'            # Metric name
```
Comment (trevorbonas) [inline correctness/mockmetheus.py:153]: These regex strings don't support [Prometheus functions](https://prometheus.io/docs/prometheus/latest/querying/functions/). For example, `sum(http_requests_total{method="GET"} @ 1609746000)`. This also means that for now we can't recreate the correctness "tests" that we [used to use](https://github.com/awslabs/amazon-timestream-connector-prometheus/blob/main/correctness/data/functions.txt). That may be fine for now if we're okay with not testing functions.

File: correctness/test_correctness.py [hunk 3]
```diff
@@ -0,0 +1,350 @@
...
+            'samples': [
+                (expected_sample_record, int(time.time() * 1000)),
+            ]
+        }
+    ]
+    write_response = mockmetheus_instance.remote_write(timeseries_data)
+    assert(write_response == 200)
+
+    # ensures data is ingested into Timestream
+    time.sleep(1)
```
Comment (trevorbonas) [inline correctness/test_correctness.py:126]: For this and the other sleeps, can network latency have any impact on whether data is finished ingesting into Timestream?

Comment (fredjoonpark) [reply correctness/test_correctness.py:126]: Technically yes, but I think it should be fair given what we can expect from their [docs](https://docs.amazonaws.cn/en_us/timestream/latest/developerguide/writes.html) ("write-to-read latency is in the sub-second range."). Cross-region requests might add additional latency, perhaps we can make the network latency sleep configurable?

Comment (trevorbonas) [reply correctness/test_correctness.py:126]: Making it configurable is a good idea. I think using a named constant, like `SLEEP_TIME`, would be good enough.

**Review by trevorbonas — APPROVED** [review]
Looks good to me once negative durations are supported and the test sleep time can be set with a named constant.

File: Dockerfile [hunk 4]
```diff
@@ -34,9 +34,6 @@ RUN go mod download
 
 COPY . .
 
-# Run unit tests for main.go and client.go
-RUN CGO_ENABLED=0 go test -tags=unit -cover -v ./timestream ./
```
Comment (forestmvey) [inline Dockerfile:38]: Why did we remove running UT with the Dockerfile?

File: correctness/mockmetheus.py [hunk 5]
```diff
@@ -0,0 +1,289 @@
+import os
```
Comment (forestmvey) [inline correctness/mockmetheus.py:1]: Add License header for new files

File: Dockerfile (hunk 4 above)
Comment (fredjoonpark) [reply Dockerfile:38]: Running UTs within Dockerfiles is generally considered bad practice.

A Dockerfile's primary purpose is to define the environment and instructions for building an image - embedding test execution conflates building and testing responsibilities which violates the principle of separation of concerns.

Docker also [builds images in layers](https://depot.dev/blog/fast-dockerfiles-theory-and-practice?utm_source=chatgpt.com), and incorporating unit tests into the dockerfile can disrupt this caching mechanism in unpredictable ways (which may lead to longer build times).

Since we already have documentation that outlines how tests should be run and we have an active CI/CD (these tests are already run in our GHA), I think this is redundant and doesn't need to be there.

File: correctness/README.md [hunk 6]
```diff
@@ -1,18 +1,21 @@
...
+1. Ensure your docker version is >= `20.10.0`, or you have `docker-compose` installed
+2. Update `docker-compose.yml` in this directory with your AWS credentials
 
 ## How to execute tests
-1. Run the following command to execute the correctness tests:
-`go test -v ./correctness`
+1. Bring up containers with `docker compose up -d`
+2. Run tests with `docker exec -it mockmetheus pytest`
+
+### Notes
```
Comment (forestmvey) [inline correctness/README.md:11]: Add a cleanup section.

File: correctness/README.md [hunk 7]
```diff
@@ -1,18 +1,21 @@
...
-2. Execute the following command to save the docker image as a compressed file and update the `version` appropriately:
-`docker save timestream-prometheus-connector-docker | gzip > timestream-prometheus-connector-docker-image-<version>.tar.gz`
+1. Ensure your docker version is >= `20.10.0`, or you have `docker-compose` installed
+2. Update `docker-compose.yml` in this directory with your AWS credentials
 
 ## How to execute tests
-1. Run the following command to execute the correctness tests:
-`go test -v ./correctness`
+1. Bring up containers with `docker compose up -d`
+2. Run tests with `docker exec -it mockmetheus pytest`
```
Comment (forestmvey) [inline correctness/README.md:9]: Can we run individual tests?

File: correctness/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,32 @@
+services:
+  mockmetheus:
+    container_name: mockmetheus
+    build: .
+    volumes:
+      - .:/app
+    working_dir: /app
+    command: ["sleep", "infinity"]
+    environment:
+      AWS_ACCESS_KEY_ID: "XXXXXXXXX"
```
Comment (forestmvey) [inline correctness/docker-compose.yml:10]: These values should be read from a `env_file` that is added to gitignore. Storing credentials in tracked files is an easy way to commit secrets.

File: correctness/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,3 @@
+FROM python:3.13-slim
+RUN pip install --no-cache-dir python-snappy requests protobuf pytest
+CMD ["sleep infinity"]
```
Comment (forestmvey) [inline correctness/Dockerfile:3]:
```suggestion
```

`sleep infinity` is called in docker-compose as well.

File: correctness/README.md [hunk 10]
```diff
@@ -1,18 +1,21 @@
...
-
-## How to build and save the docker image
-1. Execute the following command to build the docker image:
-`docker buildx build . -t timestream-prometheus-connector-docker`
-2. Execute the following command to save the docker image as a compressed file and update the `version` appropriately:
-`docker save timestream-prometheus-connector-docker | gzip > timestream-prometheus-connector-docker-image-<version>.tar.gz`
+1. Ensure your docker version is >= `20.10.0`, or you have `docker-compose` installed
+2. Update `docker-compose.yml` in this directory with your AWS credentials
 
 ## How to execute tests
```
Comment (forestmvey) [inline correctness/README.md:7]: Can we add a section how to run without docker?

File: correctness/mockmetheus.py (hunk 1 above)
Comment (forestmvey) [reply correctness/mockmetheus.py:116]: ➕

File: correctness/mockmetheus.py [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    Sample,
+)
+from google.protobuf.json_format import MessageToDict
+
+"""
+Mockmetheus is a helper class that mocks remote operations from Prometheus.
+It uses the same protobuf definitions as Prometheus to construct snappy-encoded
+payloads for both remote-read and remote-write requests and responses.
+"""
+class Mockmetheus:
```
Comment (forestmvey) [inline correctness/mockmetheus.py:23]: What is the motivation for using Python rather than Go. We are doing a lot of regex and custom parsing here for Prometheus requests rather than using Prometheus libraries.

For example we can use the `github.com/prometheus/prometheus/promql/parser` library for parsing Prometheus query strings. Using Prometheus libraries would also have the added benefit of being able to upgrade the dependencies for functional changes / bugs, and not mixing Go and Python without warrant.

**Review by forestmvey — CHANGES_REQUESTED** [review]
Would like to resolve best approach for using Go vs Python.

File: correctness/Dockerfile (hunk 9 above)
Comment (fredjoonpark) [reply correctness/Dockerfile:3]: this just makes it so it's the default command for the docker image - having it defined in the docker-compose lets us easily override it from within `docker-compose.yml`, ie. for when we want to directly run `pytest`. Let me know if you still think I should remove it from the Dockerfile

File: correctness/README.md (hunk 7 above)
Comment (fredjoonpark) [reply correctness/README.md:9]: yep! `pytest test_correctness.py::test_write_no_data`

File: correctness/mockmetheus.py (hunk 11 above)
Comment (fredjoonpark) [reply correctness/mockmetheus.py:23]: > We are doing a lot of regex and custom parsing here for Prometheus requests rather than using Prometheus libraries.

The primary goal of Mockmetheus is to control remote/read operations that a Prometheus client would typically perform. While our regex may not exactly match what Prometheus uses, we can ensure its accuracy through our test setups and the controlled environment we maintain. Our current regex and parsing approach is straightforward and should suffice unless there are specific edge cases or additional pre/post-processing steps that the Prometheus parser handles. If you think there are test cases we might be missing (cases that their parser can handle) please let me know and I will incorporate those tests to verify compatibility with our existing setup.

> not mixing Go and Python without warrant.

What's wrong with having Go and Python in one repo? The separation is clear - our connector simply exposes two REST API endpoints, and how we test against these endpoints should not matter as long as we also have the capability of verification using these same endpoints. IOW the choice of language for correctness testing does not impact the functionality of the connector itself. As long as we can effectively test and verify the endpoints, the underlying language used for tests should be flexible. However, if consistency is a priority or if there are specific reasons to standardize on Go, I'm open to rewriting the tests in Go.

Comment (forestmvey) [reply correctness/mockmetheus.py:23]: Although we can use any language for this implementation, we need some reason for language interop. If there is no benefit or clear purpose here for using Python, I would prefer we implement in Go. There is no need to write the parsing logic when we have the native parsing libraries available. Here is a small Go program which can parse a Prometheus query string, rather than introducing the required logic ourselves:

```go
package main

import (
	"fmt"
	"log"

	"github.com/prometheus/prometheus/promql/parser"
)

func main() {
	query := `scrape_duration_seconds{}[5d]`
	node, err := parser.ParseExpr(query)
	if err != nil {
		log.Fatalf("Error parsing query: %v", err)
	}
	PrintQuery(node)
}

func PrintQuery(node parser.Node) {
	switch n := node.(type) {
	case *parser.MatrixSelector:
    fmt.Println("Metric name:", n.VectorSelector)
		fmt.Println("Range:", n.Range)
	default:
		fmt.Printf("Unknown query node type: %T\n", n)
	}
}
```
